dev/buildtool/build_commands.py

```diff
@@ -0,0 +1,732 @@
+        timestring(now=test_datetime),
+        timestring(now=ref_datetime),
+        timedelta_string(ref_datetime - test_datetime))
+    return False
+
+  def _get_nebula_repository_path(self, name, force_new=False):
+    """Return the path to the repository for nebula to build.
+
+    If the repository does not exist then one will be cloned from the
+    source repository.
+
+    Nebula gets confused by our tags because it expects the Netflix ones.
+    Since we dont understand how to change nebula and dont want to invest
+    the time into it, we're just going to remove all the tags and add the
+    one we want nebula to use so it isnt confused.
+
+    Rather than modifying the original repository, we'll do all this in a
+    new one. The bonus is that we can commit and tag the repo so that
+    gradle/nebula is happy but without modifying the original repo at all
+    until we are ready to later.
+
+    Args:
+      name: [string] The name of the repository
+      force_new: [bool] If true, clear any existing repository
+    """
+    git_dir = self.source_code_manager.get_local_repository_path(name)
+    original_repo_path = os.path.abspath(git_dir)
+
+    options = self.options
+    nebula_repo_path = os.path.join(options.scratch_dir, 'build', name)
+
+    if os.path.exists(nebula_repo_path):
+      if not force_new:
+        logged_is_newer = self.__logged_is_newer.get(nebula_repo_path, False)
+        if self.__is_newer_repo(nebula_repo_path, original_repo_path,
+                                not logged_is_newer):
+          if not logged_is_newer:
+            logging.info('Reusing existing %s', nebula_repo_path)
+            self.__logged_is_newer[nebula_repo_path] = True
+          return nebula_repo_path
+
+      self.__logged_is_newer[nebula_repo_path] = False
+      logging.info('Removing old nebula repo %s', nebula_repo_path)
+      shutil.rmtree(nebula_repo_path, ignore_errors=True)
+
+    logging.debug('Determining version and build tag of %s',
+                  original_repo_path)
+    summary = self.git.collect_repository_summary(original_repo_path)
+    build_tag = '{version}-{build}'.format(
+        version=summary.version, build=options.build_number)
+
+    logging.info('Cloning %s to nebula repo %s', name, nebula_repo_path)
+    self.git.clone_repository_to_path(original_repo_path, nebula_repo_path)
+    self.git.reinit_local_repository_with_tag(
+        nebula_repo_path, build_tag,
+        'Snapshot from {path} to satisfy nebula.\n\n'
+        'commit: {commit}\ntag: {tag}\nversion: {version}'
+        .format(path=original_repo_path,
+                commit=summary.commit_id,
+                tag=summary.tag,
+                version=summary.version))
+    self.__nebula_repository_version[name] = summary.version
+    summary_path = os.path.join(
+        options.scratch_dir, name, '{name}-summary.yml'.format(name=name))
+    write_to_path(summary.to_yaml(with_commit_messages=False), summary_path)
```

Comment:
What's the intended use of this summary file?

Reply:
The summary file is persisting state about the repository (GitSummary) so it is available to other processes because the commands are run one at a time in individual processes. It is generally cheap to compute so I dont actually need it, except the nebula repository are created fresh so dont have the commit hashes or version information, never mind any tags. So the summary files provide a cache.

There is a potential vulnerability here with the different repositories in they can get out of sync. It shouldnt happen, especially if there are higher level scripts that manage it (create a new root_dir/scratch_dir or wipe the existing one)